AsyncExpiringLazy/BackgroundMonitor.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Strathweb;
+
+namespace AsyncExpiringLazy
+{
+    class BackgroundMonitor<T> : IDisposable
+    {
+        private readonly Func<Task<ExpirationMetadata<T>>> _factory;
+        private readonly Action<ExpirationMetadata<T>> _onNewItem;
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+        private readonly Task _monitorTask;
+        private int _monitorStarted;
+
+        public BackgroundMonitor(Func<Task<ExpirationMetadata<T>>> factory, Action<ExpirationMetadata<T>> onNewItem)
+        {
+            _factory = factory;
+            _onNewItem = onNewItem;
+            _monitorTask = new Task(async () => await RunMonitor().ConfigureAwait(false));
+        }
+
+        public void StartIfNotStarted()
+        {
+            if (Interlocked.CompareExchange(ref _monitorStarted, 1, 0) == 0)
+            {
+                _monitorTask.Start();
+            }
+        }
+
+        private async Task RunMonitor()
+        {
+            while (!_cts.IsCancellationRequested)
+            {
+                try
+                {
+                    var metadata = await _factory().ConfigureAwait(false);
+                    NotifyAboutNewItem(metadata);
+                    
+                    var checkAgainIn = CalculateNextIteration(metadata);
+                    Debug.WriteLine($"Sleeping the connection monitor for {checkAgainIn.TotalSeconds} seconds");
+                    await Task.Delay(checkAgainIn, _cts.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                }
+                catch (ObjectDisposedException)
+                {
+                }
+                catch (Exception e)
+                {
+                    Debug.WriteLine($"Background monitor caught unexpected error: {e}");
```

Comment:
I am not convinced this should be swallowed. 
Wouldn't this create an infinite loop if the value expires and the factory is unable to produce a new value?

Reply:
Yes, you are right. I was trying to fix a potential transient error that stops the already deployed application, which makes the background monitor unusable. 

I've changed the code to move the responsibility of handling transient errors to the owner of the factory delegate. With the proposed implementation, any unhandled exception stops the background worker and raises this exception as a result of the `Value()` method call.

Reply:
makes sense